### PR TITLE
Resolve a FIXME in prepare_AlterTableStmt_for_dispatch()

### PR DESCRIPTION
**Pull request by huansong**
GPDB has specific ALTER TABLE commands for partition tables, which
are handled differently than the other AT commands: instead of being
executed on QD first and then dispatched to QEs after all the other
commands are done, they are dispatched immediately in ATExecCmd()
and executed on QEs at the same time as QD. This results in the
possibility that AT commands could be executed in different order
in QEs than in QD.

The FIXME comment suggests to make those AT partition commands
same as other AT commands, but that could be very invasive, e.g.:

1. The AT partition commands generate a lot of subcommands that
are not AT subcommands like CREATE TABLE (see ATExecGPPartCmds()).
We cannot put them in the AT workqueue as AT subcommands.
At minimal we would need separate code to split them off the AT
workqueue, and dispatch them separately.
2. Many of the subcommands implicitly dispatch to QEs (e.g.
CreateStmt/DropStmt). For those, we need to have workarounds to
to execute them w/o dispatching, because otherwise we still could
not guarantee the same ordering of exeuction.
3. Some subcommand needs partial results from other subcommands
before executing (e.g. AT_PartSplit), which adds more complexity
in this approach.

~In this commit we opt to take a less invasive approach which is to group
the GP-specific commands together and execute them before the others.
Since the other commands' order is the same on QD and QEs, we will 
resolve the original issue.~ (this created other problem, see https://github.com/greenplum-db/gpdb/pull/14024#issuecomment-1238285142)

Easier workaround includes completely splitting the partition
subcommands off the list of ALTER TABLE subcommands, and execute
them separately (and have to be executed after all the other ALTER TABLE subcommands). 
But it is not clear if it's worth doing that:
by looking at the counter-examples (see comment for
prepare_AlterTableStmt_for_dispatch), also considering the fact
that GPDB has been doing this for a long time, the out-of-order
execution doesn't seem to cause an issue. Therefore, resolve the
FIXME as is.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/commands/tablecmds.c [hunk 1]
```diff
@@ -5071,7 +5064,7 @@ ATPrepCmd(List **wqueue, Relation rel, AlterTableCmd *cmd,
 		case AT_PartSetTemplate:
```
Comment (soumyadeep2007) [inline src/backend/commands/tablecmds.c:5064]: Let's add a comment above `case AT_PartAdd` : `GPDB specific AT commands on partitioned tables`

File: src/include/nodes/altertablenodes.h [hunk 2]
```diff
@@ -39,17 +39,18 @@
  */
 
 #define AT_PASS_UNSET			-1	/* UNSET will cause ERROR */
-#define AT_PASS_DROP			0	/* DROP (all flavors) */
-#define AT_PASS_ALTER_TYPE		1	/* ALTER COLUMN TYPE */
-#define AT_PASS_OLD_INDEX		2	/* re-add existing indexes */
-#define AT_PASS_OLD_CONSTR		3	/* re-add existing constraints */
+#define AT_PASS_GP_PART			0	/* GPDB-specific AT for partition */
```
Comment (soumyadeep2007) [inline src/include/nodes/altertablenodes.h:42]: Please update the comment:
GPDB-specific AT commands on partitions (MUST precede all passes)

File: src/include/nodes/altertablenodes.h [hunk 3]
```diff
@@ -39,17 +39,18 @@
  */
```
Comment (soumyadeep2007) [inline src/include/nodes/altertablenodes.h:39]: Lets add a comment here about the need for AT_PASS_GP_PART to be before every other pass, like you did in the commit message. That will help anyone resolving merge conflicts in this code (the person doing the resolution might be tempted, like I was, to move AT_PASS_GP_PART to the end of this list)

**Review by soumyadeep2007 — APPROVED** [review]
Great middle ground! LGTM! (just needs a little more commentary)
Thanks for the detailed commit message!

**Review by divyeshddv — APPROVED** [review]
LGTM!

**Comment by huansong** [comment]
This actually would change existing behavior of some AT commands. E.g.:

```
CREATE TABLE foo (a int) PARTITION BY range(a) (PARTITION foopar1 START(100) END(200));
ALTER TABLE foo ADD PARTITION foopar2 START(200) END(300), ALTER COLUMN a SET NOT NULL;
```

Originally after the above statements the second child wil have a NOT NULL column `a`, because the ADD PARTITION happens after the ALTER COLUMN:

```
postgres=# \d foo_1_prt_foopar2
         Table "public.foo_1_prt_foopar2"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 a      | integer |           | not null |
```

Now with this PR the column would not have NOT NULL, because (1) ADD PARTITION happens before ALTER COLUMN, and (2) ALTER COLUMN doesn't recurse into the child just created because the child didn't exist during the AT prep phase.
```
postgres=# \d foo_1_prt_foopar2
         Table "public.foo_1_prt_foopar2"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 a      | integer |           |          |
```

This was found in CI test where the pg_upgrade couldn't finish because it cannot recreate a child table that doesn't have same constraint as parent. Therefore, I think this PR would create a bug (apart from behavior change). We should find something else to tackle this problem.

**Comment by huansong** [comment]
Will opt to resolve the FIXME as is: see commit message and comment.

**Comment by ashwinstar** [comment]
So, user would be get the error message for the things that don't work properly together? And I am guessing there is no wrong behavior which happens?

**Review by ashwinstar — APPROVED** [review]

**Comment by huansong** [comment]
> So, user would be get the error message for the things that don't work properly together? And I am guessing there is no wrong behavior which happens?

No because we haven't found a case where it is an issue so we don't have an error message to throw for such a case. Basically we are maintaining the 6X behavior, but we are just adding comments to detail the potential issue/path forward more clearly.

**Comment by ashwinstar** [comment]
That was my impression, just wished to get explicit clarification. Sounds good. Merge it.

**Pull request merged**